File: Data Structures & Algorithms/permutation-string/submission-0.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.size()>s2.size())
            return false;
        vector<int> freq(26,0);
        for(char c:s1)
            freq[c-'a']++;
        vector<int> ans(26,0);
        for(int i=0;i<s1.size();i++)
            ans[s2[i]-'a']++;
        if(ans==freq)
            return true;
        int l=0;
        for(int j=s1.size();j<s2.size();j++){
            ans[s2[l]-'a']--;
            l++;
            ans[s2[j]-'a']++;
            if(ans==freq)
                return true;
        }
        return false;
    }
};

File: Data Structures & Algorithms/minimum-window-with-characters/submission-0.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        freq=defaultdict(int)
        if len(t)==0:
            return ""
        for c in t:
            freq[c]+=1
        l,r=0,0
        count=defaultdict(int)
        rcount=defaultdict(int)
        while(l<len(s) and s[l] not in freq):
            l+=1
        r=l
        ans=[-1,len(s)]
        while(r<len(s) and l<len(s)):           
            if s[r] in freq:
                if freq[s[r]]>count[s[r]]:
                    count[s[r]]+=1
                rcount[s[r]]+=1
            if count==freq:
                if ans[1]-ans[0]>=r-l:
                    ans=[l,r] 
                if rcount[s[l]]==count[s[l]]:
                    count[s[l]]-=1                    
                rcount[s[l]]-=1
                if l!=r:
                    rcount[s[r]]-=1
                    count[s[r]]-=1
                else:
                    r+=1
                l+=1
                while(l<r and s[l] not in freq):
                    l+=1
            else:
                r+=1
        if ans[0]==-1:
            return ""
        l,r=ans[0],ans[1]
        return s[l:r+1]   
            
            